modules/standard/GPU.chpl

```diff
@@ -910,21 +959,46 @@ module GPU
          writeln(Arr); // [1, 2, 3, 4, 5]
        }
   */
-  proc gpuSort(ref gpuInputArr : [] uint) {
+  proc gpuSort(ref gpuInputArr : [] ?t) {
     if !here.isGpu() then halt("gpuSort must be run on a gpu locale");
-
     if gpuInputArr.size == 0 then return;
+
+    if CHPL_GPU=="cpu" {
+      use Sort only sort;
+      sort(gpuInputArr);
+      return;
+    }
+
+    extern proc chpl_gpu_can_cub_sort(): bool;
+    if chpl_gpu_can_cub_sort() {
+      gpuCubSort(gpuInputArr);
+      return;
+    }
+
+    fallBackRadixSort(gpuInputArr);
+  }
+
+  private proc fallBackRadixSort(ref gpuInputArr : [] ?t) where isCoercible(t, uint){
     // Based on the inputArr size, get a chunkSize such that numChunks is on the order of thousands
     // TODO better heuristic here?
     var chunkSize = Math.divCeil(gpuInputArr.size, 2000);
     parallelRadixSort(gpuInputArr, bitsAtATime=8, chunkSize, noisy=false, distributed=false);
   }
 
+  private proc fallBackRadixSort(ref gpuInputArr : [] ?t) where !isCoercible(t, uint){
+    compilerError("GPU Based sorting without CUB is only supported for arrays of type uint. Please check your ROCM Version and make sure it's >= 5.0.0");
```

Comment:
This is independent of CUB, is it? Is there any chance we can fall here except with ROCm 4?

Reply:
From what I understand, the `fallBackRadixSort` is only there for ROCm 4. All CUDA versions will use the CUB sort and CPU-as-device mode will use the CPU sort. I mention CUB because the limitation of only using `uint`s arises from the absence of CUB (and because that's what the in house implementation supports).

Does this sound better:
```suggestion
    compilerError("GPU Based sorting is only supported for arrays of type uint for ROCm version <5.0.0. Upgrade your ROCm install to sort all primitive numeric types");
```

Reply:
I'm good with that.